mace/ops/pad_test.cc

```diff
@@ -156,15 +163,311 @@ void Complex(const std::vector<index_t> &input_shape,
 }  // namespace
 
 TEST_F(PadTest, ComplexFloat) {
-  Complex<float>({1, 32, 32, 4}, {0, 0, 2, 2, 1, 1, 0, 0});
-  Complex<float>({1, 31, 37, 16}, {0, 0, 2, 0, 1, 0, 0, 0});
-  Complex<float>({1, 128, 128, 32}, {0, 0, 0, 1, 0, 2, 0, 0});
+  for (int i = PadType::CONSTANT; i <= PadType::SYMMETRIC; i++) {
+    Complex<float>({1, 32, 32, 4}, {0, 0, 2, 2, 1, 1, 0, 0}, i);
+    Complex<float>({1, 31, 37, 16}, {0, 0, 2, 0, 1, 0, 0, 0}, i);
+    Complex<float>({1, 128, 128, 32}, {0, 0, 0, 1, 0, 2, 0, 0}, i);
+  }
 }
 
 TEST_F(PadTest, ComplexHalf) {
-  Complex<half>({1, 32, 32, 4}, {0, 0, 2, 2, 1, 1, 0, 0});
-  Complex<half>({1, 31, 37, 16}, {0, 0, 2, 0, 1, 0, 0, 0});
-  Complex<half>({1, 128, 128, 32}, {0, 0, 0, 1, 0, 2, 0, 0});
+  for (int i = PadType::CONSTANT; i <= PadType::SYMMETRIC; i++) {
+    Complex<half>({1, 32, 32, 4}, {0, 0, 2, 2, 1, 1, 0, 0}, i);
+    Complex<half>({1, 31, 37, 16}, {0, 0, 2, 0, 1, 0, 0, 0}, i);
+    Complex<half>({1, 128, 128, 32}, {0, 0, 0, 1, 0, 2, 0, 0}, i);
+  }
+}
+
+namespace {
+template <DeviceType D, typename T>
+void Result(const std::vector<index_t> &input_shape,
+            const std::vector<float> &input_data,
+            const std::vector<index_t> &expected_shape,
+            const std::vector<float> &expected_data,
+            const std::vector<int> &paddings,
+            const PadType pad_type) {
+  // Construct graph
+  OpsTestNet net;
+  std::string input("Input");
+  std::string t_input(input);
+  std::string output("Output");
+  std::string t_output(output);
+
+  // Add input data
+  net.AddInputFromArray<D, float>(input, input_shape, input_data);
+
+  if (D == DeviceType::CPU) {
+    t_input = "TInput";
+    t_output = "TOutput";
+    net.TransformDataFormat<DeviceType::CPU, T>(input, NHWC, t_input, NCHW);
+  }
+
+  OpDefBuilder("Pad", "PadTest")
+  .Input(t_input)
+  .Output(t_output)
+  .AddIntsArg("paddings", paddings)
+  .AddIntArg("pad_type", static_cast<int>(pad_type))
+  .AddIntArg("data_format", DataFormat::NHWC)
+  .Finalize(net.NewOperatorDef());
+
+  // Run
+  net.RunOp(D);
+
+  if (D == DeviceType::CPU) {
+    net.TransformDataFormat<DeviceType::CPU, T>(t_output, NCHW, output, NHWC);
+  }
+
+  auto actual = net.GetTensor(output.c_str());
+  auto expected = net.CreateTensor<float>(expected_shape, expected_data);
+
+  ExpectTensorNear<float>(*expected, *actual, 1e-5);
+}
+}  // namespace
+
+TEST_F(PadTest, ReflectCPU) {
```

Comment:
The organization of UTs is not very clear. The conventions are from simple UT to complex and cover as much cases as possible. My suggestions:
```
// Simple UTs: Test both CPU and GPU with specific data(float and half(GPU))
TEST_F(PadTest, Constant)()
TEST_F(PadTest, Reflect)()
TEST_F(PadTest, Symmetric)()
// Complex UTs: Test both CPU and GPU with specific data(float and half(GPU))
TEST_F(PadTest, ComplexConstant)()
TEST_F(PadTest, ComplexReflect)()
TEST_F(PadTest, ComplexSymmetric)()
```

Reply:
I've refactored tests a little, but could you clarify what was (or still is) not clear about tests organization?
Since added "ReflectCPU" and "SymmetricCPU" tests are straightforward I assume you are referring to modified "ComplexFloat", "ComplexHalf" or added "Result" (the last one now has a changed name and in a file is closer to (existing before) the simplest tests for Constant mode).
"ComplexFloat" and "ComplexHalf" - I could split those into 6 different tests (ConstantHalf, ..., SymmetricFloat) but this would result either in:
1. Code duplication (duplicated vectors input_shape and paddings) - this is a really bad idea. Unfortunately before the changes the code was already with duplicates and I didn't fix it, but duplicating it again wouldn't be correct.
or
2. Making the code more verbose - I could write 6 different tests calling a single function that runs the tests for different parameters. While it will test exactly the same thing as now, this could be considered as a better solution if the goal is to have a clear view of run tests in the report generated after testing.